source/_posts/2016-01-05-improvements-to-error-responses-in-rails-5-api-mode.markdown

```diff
@@ -0,0 +1,160 @@
+        format.json { render :show, status: :created, location: @post }
+      else
+        format.html { render :new }
+        format.json { render json: @post.errors, status: :unprocessable_entity }
+      end
+    end
+  end
+end
+``` 
+
+In Rails controllers, you have the ability to write code that can respond in different formats. By default, generated code from the scaffold command can respond on HTML or JSON formats. That is easy to see in the `create` action, but it also happens in the `index` action where the render is implicit (this action returns a list of posts and it's rendered in the appropriate format).
+
+Of course, there should be some sort of rule that determines which format is picked to respond the request. In order to understand those rules, let's now take a look at the routes generated by the scaffold.
+
+The following is the output of `bin/rake routes` in our example application:
+
+```
+   Prefix Verb   URI Pattern               Controller#Action
+    posts GET    /posts(.:format)          posts#index
+          POST   /posts(.:format)          posts#create
+ new_post GET    /posts/new(.:format)      posts#new
+edit_post GET    /posts/:id/edit(.:format) posts#edit
+     post GET    /posts/:id(.:format)      posts#show
+          PATCH  /posts/:id(.:format)      posts#update
+          PUT    /posts/:id(.:format)      posts#update
+          DELETE /posts/:id(.:format)      posts#destroy
+```
+
+These routes are what you get when you have something like `resources: :post` in your routes file. We can see that _format_ here is a dynamic segment present in every generated routes, and this is the most common way in Rails to decide which is the format used to form a response. So, in many cases the format depends on the extension in the request URL, usually appended by a client application (or client side code if it's a Rails default stack application) to indicate the expected format.
+
+Therefore, a JSON response is generated if you issue a GET request to `http:://mysite.com/posts.json` and HTML response is generated when the URL is `http:://mysite.com/posts.html`. In the case of the `index` action, it just relies on the ability of Rails to render the list of posts in the format value inferred from the route path. If the format is not specified in the request path or it's an unknown format for Rails, the response will be rendered in HTML format if it's a standard Rails application or JSON if it's configured with `config.api_only = true` (using the new API mode).
+
+By using `respond_to`, like in the 'create' action of our `PostController`, you can be specific about the content of the response depending on the specified format. In this case, if the format is not provided it will use the first format listed within the `respond_to` block. If there is not a handler for the specified format, it will raise a `ActionController::UnknownFormat` exception (and it would result in a `406 Not Acceptable` response error).
```

Comment:
This actually responds with 204 but yeah due to the bug you're working on https://github.com/rails/rails/issues/20666 it should respond with 406